text/0000-plan-flags.md

```diff
@@ -9,206 +9,238 @@
-   do not include any of these flags.
-
-## BuildPlan Changes
-  
-We propose adding `build` and `launch` fields to dependencies in the Build Plan (TOML) file. 
-  
-  Note that in this case `cache` does not make sense as this is strictly a property of a layer and is omitted.
-  
-  We propose adding the following to the Build Plan (TOML):
+  We propose the following Build Plan (TOML):
   
   ```toml
   [[provides]]
   name = "<dependency name>"
+  strategy = "<merging strategy>"
   
   [[requires]]
   name = "<dependency name>"
-  version = "<dependency version>"
-  build = false
-  launch = true
-  
-  [requires.metadata]
-  some_metadata_key = "some_metadata_value"
-  # buildpack-specific data
+  [requires.version]
+    constraint = "<dependency constraint>"
+    source = "<dependency constraint source>"
+  [requires.capabilities]
+    build = false
+    launch = false
+  [requires.modifiers]
+    some-modifer = "modifier"
   
   [[or]]
   
   [[or.provides]]
   name = "<dependency name>"
+  strategy = "<merging strategy>"
   
   [[or.requires]]
   name = "<dependency name>"
-  version = "<other dependency version>"
-  build = true
-  launch = false
-  
-  [or.requires.metadata]
-  some_other_metadata_key = "some_other_metadata_value"
-  # buildpack-specific data
-  
+  [requires.version]
+    constraint = "<other dependency constraint>"
+    source = "<other dependency constraint source>"
+  [requires.capabilities]
+    build = false
+    launch = false
+  [requires.modifiers]
+    some-other-modifer = "other modifier"
   ```
+
+This proposal removes both the `version` and `metadata` fields and replaces them with three new fields each with their own purpose.
+The `requires.version` field holds all of the version requirement information, this includes a `constraint` which can be any semver constraint
+to be used when deciding what dependency the providing buildpack should install and the `source` so that the origin of the constraint can be identified
+this will be important when discussing the merging done by the lifecycle later. `requires.capablities` hold the `build` and `launch` keys and indicate
+when during which phase buildpack specific data and dependencies are available. Finally, there is `requires.modifiers` this is meant to hold only specific
+and unique buildpack communications.
   
 ## Buildpack Plan (TOML) changes
   
-  For the same reasons, we would like to add `build` and `launch` flags to the Buildpack Plan, along with
+  For the same reasons, we would like to add the `version`, `capabilities`, and `modifiers` fields to the Buildpack Plan, along with a `versions` field and
   some additional format changes: 
 
   ```toml
   [[entries]]
   name = "<dependency name>"
-  build = false
-  launch = false
-  
-  [[entries.requires]]
-  version = "<dependency version>"
-  
-  [entries.requires.metadata]
-  # buildpack-specific data
+  [entries.version]
+    constraint = "<dependency constraint>"
```

Comment:
Why do we lose the "source" of the winning version here?

Reply:
I think the major reason comes down to the merging process, which could have a result where the constraints from 2 different sources end up in the same valid merged constraint. Say for instance we have `> 10.*` from `buildpack.yml` and `< 12.*` from `package.json` both of those would end up in the final merged constraint and there would be no good way of expressing the source.